### PR TITLE
Pf cand rescale

### PR DESCRIPTION
**Pull request by kdlong**
#### PR description:

Fix divide by zero that happens in PFCandConnector when PFCands formed from a poor displaced track are "ghosted."
* Add explicit check that energy is > 0
* Modify rescaleMomentum function to only scale three momentum, to avoid modifying the mass (and to avoid zero energy particles)

#### PR validation:

Have tested in PFValidation that no changes are seen for most events. Also confirmed that the nan events are fixed for the files reported in #39110

### CODE REVIEW

**Comment by mandrenguyen** [comment]
please test

**Comment by clacaputo** [comment]
assign pf

**Comment by clacaputo** [comment]
type pf

**Comment by kdlong** [comment]
Hi, do I need to make a back port for this to be included in the re-reco? If so, to what CMSSW version?

File: RecoParticleFlow/PFProducer/src/PFCandConnector.cc [hunk 1]
```diff
@@ -158,7 +158,8 @@ void PFCandConnector::analyseNuclearWPrim(PFCandidateCollection& pfCand,
 
   math::XYZTLorentzVectorD momentumSec;
```
Comment (perrotta) [inline RecoParticleFlow/PFProducer/src/PFCandConnector.cc:159]: I think momentumSec must be initialized, to cover the case when the `if` below is not satisfied.
Alternatively, an `else` after line 162

Comment (kdlong) [reply RecoParticleFlow/PFProducer/src/PFCandConnector.cc:159]: I did check that the default initialization is (0,0,0,0), as desired, but true that it's better to be explicit.

Comment (clacaputo) [reply RecoParticleFlow/PFProducer/src/PFCandConnector.cc:159]: > I did check that the default initialization is (0,0,0,0), as desired, but true that it's better to be explicit.

@kdlong are you planning to initialize momentumSec?

**Comment by kdlong** [comment]
please test

File: RecoParticleFlow/PFProducer/src/PFCandConnector.cc (hunk 1 above)
Comment (kdlong) [reply RecoParticleFlow/PFProducer/src/PFCandConnector.cc]: Sorry @clacaputo, it's done now. I also tried to protect against divide by zero in a few other places where it hypothetically could happen. The code would clearly benefit from cleanup but I didn't want to be too ambitious

**Comment by kdlong** [comment]
please test

**Comment by kdlong** [comment]
+pf

**Comment by kdlong** [comment]
Any additional comments here? Where should I make a back port for re-reco?

**Comment by clacaputo** [comment]
> Any additional comments here? Where should I make a back port for re-reco?

No comment from my side: code is fine and reco differences seems in line with the expectations.
About re-reco, it should be `12_4_X`

**Comment by clacaputo** [comment]
+reconstruction

- reco differences are in line with the expectation

**Comment by perrotta** [comment]
+1

**Comment by mmusich** [comment]
@kdlong @cms-sw/pf-l2 would this PR be material for a backport?

**Comment by swagata87** [comment]
Hello @cms-sw/pf-l2 

There is a ongoing validation campaign for `12_6_0_pre3_Phase2_D88`, and our photon validation expert @nancymarinelli had noticed a significant change in deltaR distribution between photon and PF neutral hadron candidates in samples with PileUp in Phase2 scenario. We are trying to find out which PR can cause this. I attach here the plot I received from Nancy. Blue is the target ie, 12_6_0_pre3_Phase2 and black is reference ie pre2. 

Do you think this PR can cause this change? Or do you know about any other PF-related PR that went between `pre2` and `pre3` that can cause this?

In the PR description, you mentioned about PFValidation. Was the validation done for Phase2 PU samples? Or only Run3 samples?

12_6_0_pre3

**Comment by slava77** [comment]
> Do you think this PR can cause this change? Or do you know about any other PF-related PR that went between `pre2` and `pre3` that can cause this?

pre2 has a broken PU mixing in the pixel IT hits.

**Comment by kdlong** [comment]
Sorry I overlooked this. Yes, the original plan was to back port this, can someone (@mmusich ?) remind me which CMSSW version it should go to?

And @swagata87 I don't think there's any way this could have such an effect, I hope the suggestion by Slava is the source.

**Comment by mmusich** [comment]
> remind me which CMSSW version it should go to?

I think you could target 12_4_X and 12_5_X, though the data-taking with 12_4_X is almost over and 12_5_X will be used only for HI tests, so the overall benefit of the backports will be limited.

**Comment by kdlong** [comment]
No objections on my side to this. Though it would be useful to have it for the re-reco: which CMSSW version will this be in?

**Comment by mmusich** [comment]
> which CMSSW version will this be in?

2022 data rereco is going to be in 12.4.x and as far as I know the release is frozen for that.

**Pull request merged**